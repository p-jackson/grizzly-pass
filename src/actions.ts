import { uniqueId } from "lodash";
import type { Project, Label, Status, TabId } from "./types";

export interface LoadDemoData {
  type: "LOAD_DEMO_DATA";
}
export const loadDemoData = (): LoadDemoData => ({
  type: "LOAD_DEMO_DATA",
});

export interface LoadProjectJsonSuccess {
  type: "LOAD_PROJECT_JSON_SUCCESS";
  data: {
    title?: string;
    projects: Project[];
    labels: Label[];
  };
}
export const loadProjectJsonSuccess = (data: {
  title?: string;
  projects: Project[];
  labels: Label[];
}): LoadProjectJsonSuccess => ({
  type: "LOAD_PROJECT_JSON_SUCCESS",
  data,
});

export interface LoadProjectJsonFailure {
  type: "LOAD_PROJECT_JSON_FAILURE";
  errorMessage: string | string[];
}
export const loadProjectJsonFailure = (
  errorMessage: string | string[],
): LoadProjectJsonFailure => ({
  type: "LOAD_PROJECT_JSON_FAILURE",
  errorMessage,
});

export interface LoadProject {
  type: "LOAD_PROJECT";
  project: Project;
}
export const loadProject = (
  title: string,
  person: string,
  time: string,
  progress: number,
  status: Status,
  labels: string[],
): LoadProject => ({
  type: "LOAD_PROJECT",
  project: {
    id: uniqueId(),
    title,
    person,
    time,
    progress,
    status,
    labels,
  },
});

export interface LoadLabel {
  type: "LOAD_LABEL";
  label: Label;
}
export const loadLabel = (title: string): LoadLabel => ({
  type: "LOAD_LABEL",
  label: {
    id: uniqueId(),
    title,
  },
});

export interface UpdateProject {
  type: "UPDATE_PROJECT";
  project: Project;
}
export const updateProject = (project: Project): UpdateProject => ({
  type: "UPDATE_PROJECT",
  project,
});

export interface SelectTab {
  type: "SELECT_TAB";
  tabId?: TabId;
}
export const selectTab = (tabId?: TabId): SelectTab => ({
  type: "SELECT_TAB",
  tabId,
});

export type Action =
  | LoadDemoData
  | LoadProjectJsonSuccess
  | LoadProjectJsonFailure
  | LoadProject
  | LoadLabel
  | UpdateProject
  | SelectTab;
