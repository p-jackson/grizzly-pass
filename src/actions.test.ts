import * as actions from "./actions";

it("creates a LoadProjectJsonSuccess action", () => {
  expect(
    actions.loadProjectJsonSuccess({
      title: undefined,
      labels: [],
      projects: []
    })
  ).toMatchSnapshot();
});

it("creates a LoadProjectJsonFailure action", () => {
  expect(actions.loadProjectJsonFailure("error")).toMatchSnapshot();
});

it("creates a LoadProjectJsonFailure action with multiple errors", () => {
  expect(
    actions.loadProjectJsonFailure(["error1", "error2"])
  ).toMatchSnapshot();
});

it("creates a LoadProject action", () => {
  expect(
    actions.loadProject("title", "person", "time", 0, "ontrack", [])
  ).toMatchSnapshot();
});

it("creates a LoadLabel action", () => {
  expect(actions.loadLabel("title")).toMatchSnapshot();
});

it("creates an UpdateProject action", () => {
  const id = "1";
  const title = "title";
  const person = "person";
  const time = "time";
  const progress = 0;
  const status = "ontrack";
  const labels = ["1", "2", "3"];
  expect(
    actions.updateProject({ id, title, person, time, progress, status, labels })
  ).toMatchSnapshot();
});

it("creates a SelectTab action", () => {
  expect(actions.selectTab("edit")).toMatchSnapshot();
});

it("can create a SelectTab action with no tab selected", () => {
  expect(actions.selectTab(undefined)).toMatchSnapshot();
});
