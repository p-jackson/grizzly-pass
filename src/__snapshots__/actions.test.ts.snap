// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`can create a SelectTab action with no tab selected 1`] = `
{
  "tabId": undefined,
  "type": "SELECT_TAB",
}
`;

exports[`creates a LoadLabel action 1`] = `
{
  "label": {
    "id": "1",
    "title": "title",
  },
  "type": "LOAD_LABEL",
}
`;

exports[`creates a LoadProject action 1`] = `
{
  "project": {
    "id": "1",
    "labels": [],
    "person": "person",
    "progress": 0,
    "status": "ontrack",
    "time": "time",
    "title": "title",
  },
  "type": "LOAD_PROJECT",
}
`;

exports[`creates a LoadProjectJsonFailure action 1`] = `
{
  "errorMessage": "error",
  "type": "LOAD_PROJECT_JSON_FAILURE",
}
`;

exports[`creates a LoadProjectJsonFailure action with multiple errors 1`] = `
{
  "errorMessage": [
    "error1",
    "error2",
  ],
  "type": "LOAD_PROJECT_JSON_FAILURE",
}
`;

exports[`creates a LoadProjectJsonSuccess action 1`] = `
{
  "data": {
    "labels": [],
    "projects": [],
    "title": undefined,
  },
  "type": "LOAD_PROJECT_JSON_SUCCESS",
}
`;

exports[`creates a SelectTab action 1`] = `
{
  "tabId": "edit",
  "type": "SELECT_TAB",
}
`;

exports[`creates an UpdateProject action 1`] = `
{
  "project": {
    "id": "1",
    "labels": [
      "1",
      "2",
      "3",
    ],
    "person": "person",
    "progress": 0,
    "status": "ontrack",
    "time": "time",
    "title": "title",
  },
  "type": "UPDATE_PROJECT",
}
`;
